dataset/src/test/scala/frameless/functions/NonAggregateFunctionsTests.scala

```diff
@@ -2,15 +2,39 @@ package frameless
 package functions
 
 import java.io.File
+
 import frameless.functions.nonAggregate._
 import org.apache.commons.io.FileUtils
 import org.apache.spark.sql.{Column, Encoder, Row, SaveMode, functions => sparkFunctions}
 import org.scalacheck.Prop._
-import org.scalacheck.{Gen, Prop}
+import org.scalacheck.{Arbitrary, Gen, Prop}
 
 class NonAggregateFunctionsTests extends TypedDatasetSuite {
   val testTempFiles = "target/testoutput"
 
+
+  object NonNegativeGenerators {
```

Comment:
How is frameless failing compared to vanilla Spark in case of negative value? I'm a bit nervous about having that many custom generator in property tests, do you think we would get it more fine grained?

It would be nice to document exactly what bounds we expect on Doubles and have the tests match exactly that. Ideally I would also like to have some evidence in these tests that frameless fails reasonably nicely when these contracts are violated...

Reply:
it's not the case that framless is failing and spark do not. It's just because of log function which takes nonnegative numbers since giving him for example 0 base caused runtime exception.

Reply:
> it's not the case that framless is failing and spark do not.

It that case I think I would prefer if we could avoid using custom number generators. Instead, the tests should assert that frameless and Spark behave the same for both correct and incorrect inputs.

Reply:
actually I am a bit confused, because I don't know how you would achieve such result. If I did custom prop[Int] it would cause runtime exception if any of the generated number wasn't positive. So should I than try catch it somehow instead of providing positive numbers? Could you give me a hint how you would like to test of such function look?

Reply:
Yes, the idea would be generate uniform numbers, catch exceptions in the test code, and make sure Spark and Frameless agree on everything.

In pseudocode:

```scala
forAll[Int] { i =>
  val expected = scala.util.Try(spark.log(i))
  val result = scala.util.Try(frameless.log(i))
  (expected, result) match {
    case (Failure(e1), Failure(e2)) => // Both fail, check that e1 & e2 are the same
    case (Success(s1), Success(s2)) => // Check that s1 & s2 are the same
    case => ??? // Test failure
  }
}
```